File: backend/app/domains/flood/funding_gap.py
"""PMDRF cash vs RDNA recovery need — Atlas arithmetic, not an official shortfall.

The two inputs are separately reviewed figures that Atlas already publishes.
This module only converts units and subtracts. It never invents a MoF or NPC
shortfall claim, and it never adds pledges or in-kind onto the cash bar.
"""

from __future__ import annotations

from typing import Any

NPR_PER_CRORE = 10_000_000

CAVEAT_EN = (
    "Atlas arithmetic — not an official MoF or NPC shortfall. "
    "PM Disaster Relief Fund bank stock versus NPC–NDRRMA preliminary RDNA "
    "recovery need. Scopes differ; PhonePe, foreign pledges and in-kind are "
    "not in the cash bar."
)
CAVEAT_NE = (
    "एट्लसको अंकगणित — आधिकारिक MoF वा NPC अभाव होइन। "
    "प्रधानमन्त्री दैवी प्रकोप उद्धार कोषको बैंक मौज्दात विरुद्ध "
    "NPC–NDRRMA प्रारम्भिक RDNA पुनर्प्राप्ति आवश्यकता। "
    "क्षेत्र फरक; फोनपे, वैदेशिक घोषणा र जिन्सी यो नगद बारमा छैनन्।"
)


def npr_to_crore(npr: float | int | None) -> float | None:
    if npr is None:
        return None
    return round(float(npr) / NPR_PER_CRORE, 2)


def _pm_fund_npr(relief_received: dict[str, Any] | None) -> float | None:
    if not relief_received:
        return None
    pm = next(
        (h for h in (relief_received.get("headline") or []) if h.get("id") == "pm-fund"),
        None,
    )
    if not pm or pm.get("value") is None:
        return None
    return float(pm["value"])


def _rdna_recovery_cr(damage: dict[str, Any] | None) -> float | None:
    rdna = (damage or {}).get("rdna") or {}
    recovery = next(
        (h for h in (rdna.get("headline") or []) if h.get("id") == "recovery"),
        None,
    )
    if recovery and recovery.get("value_cr") is not None:
        return float(recovery["value_cr"])
    total = next((r for r in (rdna.get("rows") or []) if r.get("id") == "total"), None)
    if total and total.get("recovery_cr") is not None:
        return float(total["recovery_cr"])
    return None


def _rdna_effects_cr(damage: dict[str, Any] | None) -> float | None:
    rdna = (damage or {}).get("rdna") or {}
    effects = next(
        (h for h in (rdna.get("headline") or []) if h.get("id") == "effects"),
        None,
    )
    if effects and effects.get("value_cr") is not None:
        return float(effects["value_cr"])
    total = next((r for r in (rdna.get("rows") or []) if r.get("id") == "total"), None)
    if total and total.get("effects_cr") is not None:
        return float(total["effects_cr"])
    return None


def compute_funding_gap(
    *,
    relief_received: dict[str, Any] | None,
    damage: dict[str, Any] | None,
) -> dict[str, Any] | None:
    """Return crore bars for received · recovery · difference, or None if either side is missing."""
    received_cr = npr_to_crore(_pm_fund_npr(relief_received))
    recovery_cr = _rdna_recovery_cr(damage)
    if received_cr is None or recovery_cr is None:
        return None

    effects_cr = _rdna_effects_cr(damage)
    gap_cr = round(recovery_cr - received_cr, 2)
    received = relief_received or {}
    rdna = (damage or {}).get("rdna") or {}

    return {
        "unit": "npr_cr",
        "received_cr": received_cr,
        "recovery_cr": recovery_cr,
        "effects_cr": effects_cr,
        "gap_cr": gap_cr,
        "received_as_of": received.get("as_of_label_en") or received.get("as_of"),
        "received_as_of_ne": received.get("as_of_label_ne") or received.get("as_of_label_en"),
        "rdna_as_of": rdna.get("as_of_label_en") or (damage or {}).get("as_of_label_en"),
        "rdna_as_of_ne": rdna.get("as_of_label_ne")
        or rdna.get("as_of_label_en")
        or (damage or {}).get("as_of_label_ne"),
        "caveat_en": CAVEAT_EN,
        "caveat_ne": CAVEAT_NE,
        "bars": [
            {
                "id": "received",
                "label_en": "PM fund cash",
                "label_ne": "प्रधानमन्त्री कोष नगद",
                "value_cr": received_cr,
            },
            {
                "id": "recovery",
                "label_en": "RDNA recovery need",
                "label_ne": "RDNA पुनर्प्राप्ति आवश्यकता",
                "value_cr": recovery_cr,
            },
            {
                "id": "gap",
                "label_en": "Difference",
                "label_ne": "अन्तर",
                "value_cr": gap_cr,
            },
        ],
    }


def funding_gap_from_snap(snap: dict[str, Any]) -> dict[str, Any] | None:
    return compute_funding_gap(
        relief_received=snap.get("reliefReceived"),
        damage=snap.get("damage"),
    )


def funding_gap_chart_view(snap: dict[str, Any]) -> dict[str, Any] | None:
    gap = funding_gap_from_snap(snap)
    if not gap:
        return None
    return {
        "chart": "funding_gap",
        "unit": gap["unit"],
        "bars": gap["bars"],
        "caveat_en": gap["caveat_en"],
        "caveat_ne": gap["caveat_ne"],
    }

File: frontend/src/lib/funding-gap.ts
/** PM fund available vs RDNA recovery need — mirrors backend/app/domains/flood/funding_gap.py */

export const NPR_PER_CRORE = 10_000_000;

export const FUNDING_GAP_CAVEAT = {
  en:
    'Atlas arithmetic — not an official MoF or NPC shortfall. PM Disaster Relief Fund available cash (NPR bank stock + USD accounts at the published rate) versus NPC–NDRRMA preliminary RDNA recovery need. Foreign pledges outside the fund table, PhonePe, World Bank and IFRC are not in the cash bar.',
  ne:
    'एट्लसको अंकगणित — आधिकारिक MoF वा NPC अभाव होइन। प्रधानमन्त्री दैवी प्रकोप उद्धार कोषको उपलब्ध नगद (NPR बैंक मौज्दात + प्रकाशित दरमा USD खाता) विरुद्ध NPC–NDRRMA प्रारम्भिक RDNA पुनर्प्राप्ति आवश्यकता। कोष तालिकाबाहिरका वैदेशिक घोषणा, फोनपे, विश्व बैंक र आईएफआरसी यो नगद बारमा छैनन्।',
} as const;

export interface FundingGapBar {
  id: 'received' | 'recovery' | 'gap';
  label_en: string;
  label_ne: string;
  value_cr: number;
}

export interface FundingGap {
  unit: 'npr_cr';
  received_cr: number;
  npr_stock_cr: number | null;
  includes_usd_accounts: boolean;
  recovery_cr: number;
  effects_cr: number | null;
  gap_cr: number;
  received_as_of: string | null;
  received_as_of_ne: string | null;
  rdna_as_of: string | null;
  rdna_as_of_ne: string | null;
  bars: FundingGapBar[];
  caveat_en: string;
  caveat_ne: string;
}

type ReliefSlice = {
  as_of?: string;
  as_of_label_en?: string;
  as_of_label_ne?: string;
  total_available_npr?: number | null;
  headline?: Array<{ id?: string; value?: number | null }>;
  breakdowns?: Array<{
    id?: string;
    aside?: Array<{
      label_en?: string;
      value?: number | null;
      unit_en?: string;
    }>;
  }>;
};

export function nprToCrore(npr: number | null | undefined): number | null {
  if (npr == null) return null;
  return Math.round((npr / NPR_PER_CRORE) * 100) / 100;
}

function pmFundNprStock(received?: ReliefSlice | null): number | null {
  const pm = (received?.headline || []).find(h => h.id === 'pm-fund');
  return pm?.value != null ? Number(pm.value) : null;
}

function pmFundTotalAvailableNpr(received?: ReliefSlice | null): number | null {
  if (received?.total_available_npr != null) return Number(received.total_available_npr);
  for (const group of received?.breakdowns || []) {
    if (group.id !== 'pm-fund') continue;
    for (const row of group.aside || []) {
      const label = (row.label_en || '').toLowerCase();
      if (!label.includes('total available')) continue;
      if (row.value == null) continue;
      return Number(row.value);
    }
  }
  return null;
}

export function computeFundingGap(input: {
  reliefReceived?: ReliefSlice | null;
  damage?: {
    as_of_label_en?: string;
    as_of_label_ne?: string;
    rdna?: {
      as_of_label_en?: string;
      as_of_label_ne?: string;
      headline?: Array<{ id?: string; value_cr?: number | null }>;
      rows?: Array<{
        id?: string;
        effects_cr?: number | null;
        recovery_cr?: number | null;
      }>;
    };
  } | null;
}): FundingGap | null {
  const received = input.reliefReceived;
  const rdna = input.damage?.rdna;
  const stockNpr = pmFundNprStock(received);
  const availableNpr = pmFundTotalAvailableNpr(received) ?? stockNpr;
  const receivedCr = nprToCrore(availableNpr);

  const recoveryHeadline = (rdna?.headline || []).find(h => h.id === 'recovery');
  const effectsHeadline = (rdna?.headline || []).find(h => h.id === 'effects');
  const total = (rdna?.rows || []).find(r => r.id === 'total');
  const recoveryCr =
    recoveryHeadline?.value_cr != null
      ? Number(recoveryHeadline.value_cr)
      : total?.recovery_cr != null
        ? Number(total.recovery_cr)
        : null;
  const effectsCr =
    effectsHeadline?.value_cr != null
      ? Number(effectsHeadline.value_cr)
      : total?.effects_cr != null
        ? Number(total.effects_cr)
        : null;

  if (receivedCr == null || recoveryCr == null) return null;

  const gapCr = Math.round((recoveryCr - receivedCr) * 100) / 100;
  const stockCr = nprToCrore(stockNpr);
  const includesUsd =
    stockNpr != null && availableNpr != null && Math.abs(availableNpr - stockNpr) > 1;

  return {
    unit: 'npr_cr',
    received_cr: receivedCr,
    npr_stock_cr: stockCr,
    includes_usd_accounts: includesUsd,
    recovery_cr: recoveryCr,
    effects_cr: effectsCr,
    gap_cr: gapCr,
    received_as_of: received?.as_of_label_en || received?.as_of || null,
    received_as_of_ne: received?.as_of_label_ne || received?.as_of_label_en || null,
    rdna_as_of: rdna?.as_of_label_en || input.damage?.as_of_label_en || null,
    rdna_as_of_ne:
      rdna?.as_of_label_ne || rdna?.as_of_label_en || input.damage?.as_of_label_ne || null,
    caveat_en: FUNDING_GAP_CAVEAT.en,
    caveat_ne: FUNDING_GAP_CAVEAT.ne,
    bars: [
      {
        id: 'received',
        label_en: 'PM fund available',
        label_ne: 'प्रधानमन्त्री कोष उपलब्ध',
        value_cr: receivedCr,
      },
      {
        id: 'recovery',
        label_en: 'RDNA recovery need',
        label_ne: 'RDNA पुनर्प्राप्ति आवश्यकता',
        value_cr: recoveryCr,
      },
      {
        id: 'gap',
        label_en: 'Difference',
        label_ne: 'अन्तर',
        value_cr: gapCr,
      },
    ],
  };
}

export function formatCrore(value: number): string {
  return value.toLocaleString(undefined, {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  });
}
